Realistic vehicle wheel speed Sandcastle example

OmarShehata opened the pull request:
This is for https://github.com/AnalyticalGraphicsInc/cesium/issues/7295.

I've piggybacked off of the Multi-Part CZML example because it was the most relevant one. I didn't want to add a new Sandcastle example because I'm not sure that it's relevant to enough users, and I think too many Sandcastle examples make it hard to find things. I can see this being its own example if it's all about manipulating animations though. 

The only thing we "lose" here is using the ground vehicle instead of the Cesium milk truck. I used it because it's easier to see the wheels spin on it. 

In order to do this I had to expose an option to control animation speed through the Entity API. Initially I wanted to expose either the `Model` object or at least the `ModelAnimationCollection` through the Entity API, but that means it starts out undefined, and only gets set after the update step, and ModelGraphics doesn't seem to have a notion of "ready". 

The one thing I don't like is that if you constantly update the `speedup` property the animation keeps skipping. That's because the scene time controls the absolute position of the animation, as opposed to it just controlling how big the step is. I was going to refactor so `speedup` just scaled the delta (that way, negative speeds are allowed and the animation just goes backwards) but that approach means if you jump to a specific time the model animation may not always be at the same spot in that time. So I think this structure is intentional, we just have to not update it every frame. 

@lilleyse do you want to take a first review on it?
 
TODO
* [ ] Update CHANGES.md with the new `speedupAnimations` property.

cesium-concierge commented:
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/vehicle-wheel-speed/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

OmarShehata commented:
What's cool about it is that, you can see how the wheels slow/speed up based on moving the time, but the wheel speed is computed based on the absolute distance the entity traverses. So if you update the CZML to make it move faster/slower, or if the entity is just moved faster (as in, if it weren't controlled by the CZML) it would still have the correct speed. 

It's also based on the wheel's circumference, so to apply this to another model you'd just need to update the circumference. 

http://localhost:8080/Apps/Sandcastle/index.html?src=Multi-part%20CZML.html

![wheels1](https://user-images.githubusercontent.com/1711126/49387872-df13c980-f6f0-11e8-9a6e-ae902911f42a.gif)

![wheels2](https://user-images.githubusercontent.com/1711126/49387877-e1762380-f6f0-11e8-9471-d6a7d004ed42.gif)

hpinkos commented on Source/DataSources/ModelGraphics.js:
```diff
@@ -190,6 +191,14 @@ define([
+         * @type {Property}
+         * @default 1.0
+         */
+        speedupAnimations : createPropertyDescriptor('speedupAnimations'),
```
This should probably be called `animationSpeedMultiplier` or something like that

hpinkos replied:
Ah, I see it's already called `speedup` in `ModelAnimation`.  Is that a common term in the animation world?  We might want to rename it otherwise.  `speedup` has no meaning to me, but I know intuitively what a `multiplier` is.

OmarShehata replied:
Yeah I wasn't happy with `speedup` either, but I guess it would be a breaking change to rename that now. 

The original intent might have been to use this variable to speed up animations but multiplier makes more sense since you can slow it down too.

hpinkos replied:
Cool.  Can you open a separate PR deprecating `speedup` and replacing it with `animationMultiplier`?  It's not worth adding new weird attribute names because we had a weird name for an old one.

hpinkos commented on Source/DataSources/ModelVisualizer.js:
```diff
@@ -176,6 +179,16 @@ define([
+                // Update the speedup property
+                if (modelData.animationsRunning && modelData.speedup !== speedup) {
+                    var animationLength = model.activeAnimations.length;
+                    for (var animationIndex = 0; animationIndex < animationLength; ++animationIndex) {
```
Rename `animationIndex` to `i` for simplicity and readability

OmarShehata replied:
I was following the style a few lines below here: https://github.com/AnalyticalGraphicsInc/cesium/blob/7b36978f16d55ae7c290047e79b24ff212714474/Source/DataSources/ModelVisualizer.js#L197-L198

I think `i` is already defined in the outer loop so I'll name it `j`.

hpinkos commented on Source/Scene/ModelAnimation.js:
```diff
@@ -200,13 +205,19 @@ define([
+            },
+            set : function(value) {
+                if (defined(value) && (value <= 0.0)) {
+                    throw new DeveloperError('speedup must be greater than zero.');
```
use `Check.typeOf.number.greaterThan`

OmarShehata commented:
Thank you both for the feedback! While we're reviewing this, we might want to consider what all we need to expose through the Entity API for models. Here's a forum discussion where a user needs access to the `ModelAnimation` which I think currently there'd be no way to do it through entities:

https://groups.google.com/d/msg/cesium-dev/4wVIah9ZAuk/A39Gi6POBAAJ

Unless the answer is "Just use a primitive for that one". Which might be okay if it's easy to position it, but I haven't tried it so I don't remember how positioning works there.

hpinkos commented:
@OmarShehata see https://github.com/AnalyticalGraphicsInc/cesium/issues/7378#issuecomment-444120667

pjcozzi commented:
Please bump to me when ready for a quick pass.

OmarShehata commented:
@pjcozzi the actual example is ready to look at if you want to review that now (regarding things like, does the Sandcastle show this feature well enough? is it ok to put it in this Sandcastle or should I separate it out into its own?)

I'm just waiting on feedback about the Entity API changes so the implementation/property names might change.

impactblue573 commented:
Great work guys, this is a highly valuable addition!

mramato commented:
Some high level feedback, I'll dig in deeper once we hash out the details.

1. I don't think globbing on to the existing example is a good idea.  One of Sandcastle's biggest problems are examples that try and do too much.  I would separate this out into it's own example with a single simple CZML or even just use the Entity API directly to construct a simple path.
2. This only works because it is speeding up all animations on the model.  So this isn't any good for models with animations other than wheels.  That seems to make it less useful.  Why isn't this using the built in Entity NodeTransformation stuff to rotate each wheel? (For example this does it through CZML but should still paint the picture: https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/?src=CZML%20Model%20-%20Node%20Transformations.html)
3. The computation should no be happening in the onTick, it should be a Callback or other Entity Property that does the computation only when needed.  This will also make sure it supports jumping around on the timeline and other stuff.
4. Rather than perform finite differencing yourself, we should create and use a VelocityVectorProperty property as part of your CallbackProperty which will simplify the code.

OmarShehata commented:
That all seems like great feedback @mramato. Except:

> Why isn't this using the built in Entity NodeTransformation stuff to rotate each wheel? 

It seemed like a far simpler workflow for me to have a car with animated wheels, and just apply a speed multiplier than hardcode quaternions into my data. The fact that it is applied to all animations instead of just the wheels is a limitation of the Entity API that I think we should ameliorate. 

Plus that way you can easily switch it out with another animated car model and not have to go back and change all the names of the nodes in the CZML.

OmarShehata commented:
Ok it is now a new, separate Sandcastle example, and I must say I am quite pleased with how easy it was using the Entity's properties and how clean everything's turned out! 

This was probably the most enjoyable experience I've had working with CesiumJS once @mramato illuminated me to the power and joy of the composability of Entity properties. All I really had to do was create a list of positions. The orientation of the vehicle is automatically computed from those at any given moment, and the animation multiplier just uses the dynamic velocity vector property to figure out the relative speed to play at. 

Best of all? It all works perfectly with the timeline. If you play the animation at 2x, the actual vehicle speed in meters per second does **not** change, but the animation speed automatically does, so it's all still accurate and perfect. It's straightforward, it's flexible, it's dynamic, it's beautiful.

The only thing I didn't implement is manually rotating wheels with NodeTransformation because I still maintain that this is a better workflow and not being able to control individual animations with the Entity API is a separate issue.

hpinkos commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -397,8 +398,9 @@ define([
-            if (!defined(scheduledAnimation._duration)) {
+            if (!defined(scheduledAnimation._duration) || scheduledAnimation._multiplierChanged) {
                 scheduledAnimation._duration = runtimeAnimation.stopTime * (1.0 / scheduledAnimation.multiplier);
+                scheduledAnimation._multiplierChanged = false;
```
Should these be a separate if blocks?  Why do we need to update both `duration` and `multiplierChanged`?

hpinkos replied:
Is `multiplierChanged` actually used anywhere?  I can't find it

OmarShehata replied:
See answer below https://github.com/AnalyticalGraphicsInc/cesium/pull/7361#discussion_r244353957

hpinkos commented on Source/Scene/ModelAnimation.js:
```diff
@@ -208,13 +211,18 @@ define([
                 return this._multiplier;
+            },
+            set : function(multiplier) {
+                Check.typeOf.number.greaterThan('multiplier', multiplier, 0);
```
Always put `Check` in pragmas

```js
//>>includeStart('debug', pragmas.debug);
Check.typeOf......
//>>includeEnd('debug');
```

hpinkos commented on Source/Scene/ModelAnimation.js:
```diff
@@ -118,6 +120,7 @@ define([
         // Set during animation update
         this._computedStartTime = undefined;
         this._duration = undefined;
+        this._speedupChanged = false;
```
What's this for?

hpinkos commented:
@OmarShehata specs?

Everything else looks good.  As soon as you add specs for the new properties this should be good to go.

OmarShehata commented:
I've updated the specs - in doing so I realized the new multiplier wasn't included in the clone and merge functions, so I fixed that as well. The other test will make sure the entity API does correctly modify the animation object. Tests for whether `multiplier` does the right thing in terms of the animation are covered at the primitive level in ModelSpec.

OmarShehata commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -64,6 +64,7 @@ define([
 
         this._model = model;
         this._scheduledAnimations = [];
+        this._multiplierChanged = false;
```
It is. See https://github.com/AnalyticalGraphicsInc/cesium/pull/7361#discussion_r244353957

hpinkos replied:
You have `multiplierChanged` on `ModelAnimation` and I see that that's being used.  But this is on `ModelAnimationCollection` and I don't see that being set or accessed anywhere.

OmarShehata replied:
Ah, yes, you're right. My bad!

OmarShehata commented:
Thanks for catching those @hpinkos ! It should be good now. Although the build failed with a bunch of timeouts, I believe it's a fluke so I restarted the build. I'll ping you again when it passes.

OmarShehata commented:
@hpinkos all good now!

mramato commented:
@OmarShehata, this and all of your other PRs are on my 2018 catch-up list for today.

mramato commented:
Would it make sense to show a simple mph/kmh label so that you know how fast the truck is traveling?  When I run the example without reading anything, it's not immediately clear what I'm looking at.  Might also be good to have the truck come to a complete stop halfway through or something..

mramato commented on Apps/Sandcastle/gallery/Animation Speed.html:
```diff
@@ -0,0 +1,115 @@
+    position.addSample(time, location);
+}
+
+// A velocity vector property will give us the entity's speed and direction at any given time.
```
I would move these variables to right before the function that uses them, so that it's more clear to the reader.

mramato commented on Apps/Sandcastle/gallery/Animation Speed.html:
```diff
@@ -0,0 +1,115 @@
+    return multiplier;
+}, false);
+
+function rotationSpeedFromVelocity(velocity, wheelRadius) {
```
Since this function is only called once, I would inline it so simplify the code further.

mramato commented on Apps/Sandcastle/gallery/Animation Speed.html:
```diff
@@ -0,0 +1,115 @@
+    // animationMultiplier can't be 0.
+    multiplier = Math.max(Math.abs(multiplier), 0.001);
+    // If it gets too fast it doesn't look great so we create a maximum here.
+    multiplier = Math.min(multiplier, 15);
```
Why?  If you have a video of tire, sometimes it looks weird due to fps, why would we change that here?  Let's model the real world.

mramato commented on Apps/Sandcastle/gallery/Animation Speed.html:
```diff
@@ -0,0 +1,115 @@
+    velocityVectorProperty.getValue(time, velocityVector);
+
+    var rotationsPerSecond = rotationSpeedFromVelocity(Cesium.Cartesian3.magnitude(velocityVector), 0.52);
+    // Make the animation play at that speed, given that this glTF wheel animation makes almost 2 rotations per second.
```
I think we need a clearer comment here.  It's not obvious to a reader that we are animating everything about the model, and that this model just happens to have 4 wheel animations, instead of animating each wheel directly.  Also, please write up an issue with whatever is blocking you from animating the wheels directly and any thoughts on how to fix it so that we can revisit in the future.

mramato commented on Apps/Sandcastle/gallery/Animation Speed.html:
```diff
@@ -0,0 +1,115 @@
+
+// Create a path from a set of positions for our vehicle.
+var position = new Cesium.SampledPositionProperty();
+// These numbers were obtained by starting with 3 Cartesian3 locations, and then using
```
Why not just lerp between 2 points and use non-uniform times?  This will cause the car to travel and different speeds and simplify the code to two points (start/end cartographics).

mramato commented on Source/DataSources/ModelVisualizer.js:
```diff
@@ -176,6 +179,15 @@ define([
                 }
 
+                if (modelData.animationsRunning && modelData.multiplier !== multiplier) {
+                    var animationLength = model.activeAnimations.length;
```
What exactly is an `activeAnimation` can an animation become inactive?  And if that happens will it's multiplier setting become out of date?

mramato commented:
Just those few comments.  This is looking much cleaner.  Thanks again @OmarShehata!

OmarShehata commented:
I refactored the example to use nodeTransformations because I ran into an issue when the vehicle accelerates. The animation is going to be very jumpy because triggering a change in the `multiplier` will change the duration of the animation, and thus where it is in its rotation. This is intentional because it's time-based instead of frame-based, but it makes it unusable for this case.

I think it's now cleaner and even more robust. The trick was just to think of everything as data mapped over time, as opposed to frame-based logic. This should resolve all the issues you brought up @mramato.

I renamed it to `Time Dynamic Wheels` because it's no longer using animations as the speed up mechanic. I think this will be a very useful example of using dynamic node transformations.

mramato commented:
Nice!  Sounds awesome.  I'll try to take a look soon and get this merged this week.

OmarShehata commented:
I totally forgot about this! Bumping it again @mramato . As long as it gets merged before the next release I'll be happy.

mramato commented:
Me too!  Thanks for the reminder.

mramato commented:
Definitely looks smaller and cleaner than the original example, nice job!  Just two last comments:

1. Do we need to expose `animationsMultiplier` at all now?  Looks like the Sandcastle example is no longer using it, so isn't this PR just a 1 file change now (the new example)?  We try not to expose features "just because" since it becomes a maintenance tail, so unless there's a good reason to keep it, I would remove everything except what's needed by the example (we can always bring it out of the history if needed)

2. While it makes sense to pre-compute the position/velocity (because these are usually provided externally) wouldn't it be a better example if we made `wheelAngleProperty` a CallbackProperty that was computed on the fly?  This would allow our users to easily copy/paste/re-use it in their own applications (in fact, we might even eventually make `WheelAngleFromVelocity` a standard property if it's popular enough.

mramato commented on Apps/Sandcastle/gallery/Time Dynamic Wheels.html:
```diff
@@ -0,0 +1,129 @@
+    orientation : new Cesium.VelocityOrientationProperty(position), // Automatically set the vehicle's orientation to the direction it's facing.
+    model : {
+        uri : '../../../../Apps/SampleData/models/GroundVehicle/GroundVehicle.glb',
+        minimumPixelSize : 128,
```
This and maximumScale just add news to the demo, since we automatically track the entity, there's no reason to set these, users will see it no problem.

OmarShehata commented:
@mramato thank you for the thorough review and for pushing to make this example as robust and re-usable as possible!

Should be ready now. I'm not sure why `Source/DataSources/ModelGraphics.js` is appearing in the Files Changed for me, given that this change was made and already merged in https://github.com/AnalyticalGraphicsInc/cesium/pull/7464

mramato commented:
It will show up if you haven't actually merged upstream master into your branch. Either way, it's not a big deal.  Thanks again!

cesium-concierge commented:
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/d/msg/cesium-dev/4wVIah9ZAuk/A39Gi6POBAAJ

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/pull/7361 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:
